jdaviz/app.py

```diff
@@ -1570,6 +1571,61 @@ def get_viewer_reference_names(self):
         # Cannot sort because of None
         return [self._viewer_item_by_id(vid).get('reference') for vid in self._viewer_store]
 
+    def update_viewer_reference_name(
+        self, old_reference, new_reference,
+        old_reference_helper_attr=None, update_id=False
+    ):
+        """
+        Update viewer reference names.
+
+        Viewer IDs will not be changed unless `update_id` is True.
+
+        Parameters
+        ----------
+        old_reference : str
+            The viewer reference name to be changed
+        new_reference : str
+            The viewer reference name to use instead of ``old_reference``
+        old_reference_helper_attr : str, optional
+            If one exists, update the helper's attribute `old_reference_helper_attr`
+            with the new viewer reference name. Default is None.
+        update_id : bool, optional
+            If True, update the viewer IDs as well as the viewer reference names.
+        """
+        if new_reference in self.get_viewer_reference_names():
+            raise ValueError(f"viewer with reference='{new_reference}' already exists")
+
+        # update the viewer item's reference name
+        viewer_item = self._get_viewer_item(old_reference)
+        viewer_item['reference'] = new_reference
+
+        if viewer_item['name'] == old_reference:
+            viewer_item['name'] = new_reference
+
+        # optionally update the viewer IDs:
+        if update_id and viewer_item['id'] == old_reference:
+            # update the id as well
+            old_id = viewer_item['id']
+            viewer_item['id'] = new_reference
+            self._viewer_store[new_reference] = self._viewer_store.pop(old_id)
+            self.state.viewer_icons[new_reference] = self.state.viewer_icons.pop(old_id)
+
+        # Update the viewer name attributes on the helper. First check if
+        # an attr for this viewer reference name is stored on the helper:
+        if old_reference_helper_attr is None:
+            old_viewer_ref_attrs = [
+                attr for attr in dir(self._jdaviz_helper)
+                if old_reference.replace('-', '_') in attr
+            ]
```

Comment:
if I'm following this correctly, won't this only work on the _first_ rename?  Would something like the following work and maybe allow us to remove the ability to pass as a kwarg?

```
if attr.startswith('_default') and getattr(self._jdaviz_helper, attr) == old_reference
```

Reply:
Good idea. I generalized this a bit in [196f91e](https://github.com/spacetelescope/jdaviz/pull/2338/commits/196f91e3094c73f6f164d20acc45ceefde806988).